### PR TITLE
x509: add missing X509 dup functions

### PR DESCRIPTION
**Pull request by mspncp**
_Note: Some of these missing 'duplicators' are urgently needed by me for backporting some company code from 1.0.2 to 1.1.1._

_That's why backporting this change to 1.1.1 (as a separate pull request) is my primary target. Since the backport is based on #9347, it will depend on the OMC decision about the constification policy._

_If backporting will not be possible, well, then I'll have to live with my downstream patches...._

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
> That's why backporting this change to 1.1.1 (as a separate pull request) is my primary target. Since the backport is based on #9347, it will depend on the OMC decision about the constification policy.

We've generally allowed constification backports where it doesn't break API/ABI compat.

**Comment by mspncp** [comment]
> > That's why backporting this change to 1.1.1 (as a separate pull request) is my primary target. Since the backport is based on #9347, it will depend on the OMC decision about the constification policy.
> 
> We've generally allowed constification backports where it doesn't break API/ABI compat.

Actually, this pull request here is a 'missing accessor' pull request. Number #9347 is the 'constification' pull request which is currently being discussed. Because we learned that in some specific cases, constifications can cause compiler warnings or even errors. Please take a look at the discussion at https://github.com/openssl/openssl/pull/9347#pullrequestreview-260765116.

**Comment by mspncp** [comment]
Kicking Travis...

File: include/openssl/x509.h [hunk 1]
```diff
@@ -466,12 +466,13 @@ int i2d_PUBKEY_bio(BIO *bp, const EVP_PKEY *pkey);
...
+DECLARE_ASN1_DUP_FUNCTION(X509_ALGOR)
 DECLARE_ASN1_DUP_FUNCTION(X509_ATTRIBUTE)
-DECLARE_ASN1_DUP_FUNCTION(X509_EXTENSION)
 DECLARE_ASN1_DUP_FUNCTION(X509_CRL)
-DECLARE_ASN1_DUP_FUNCTION(X509_REVOKED)
+DECLARE_ASN1_DUP_FUNCTION(X509_EXTENSION)
+DECLARE_ASN1_DUP_FUNCTION(X509_PUBKEY)
 DECLARE_ASN1_DUP_FUNCTION(X509_REQ)
-DECLARE_ASN1_DUP_FUNCTION(X509_ALGOR)
+DECLARE_ASN1_DUP_FUNCTION(X509_REVOKED)
```
Comment (slontis) [inline include/openssl/x509.h]: Maybe some unit tests would be good - coverage is already bad enough :)

Comment (mspncp) [reply include/openssl/x509.h]: Are you serious? Or is this just your tit-for-tat response to my last overly pedantic comments? ;-)

Comment (slontis) [reply include/openssl/x509.h]: up to you - at some point coverage will be looked at.. maybe..

Comment (mspncp) [reply include/openssl/x509.h]: I agree with you that the ASN1 parser could need some more covering by unit tests.  But I am not the person who would be able to implement them.  Also, I am already punished enough with all those missing accessors and const incorrectnesses encountered while porting code from 1.0.2 to 1.1.1.

**Comment by mspncp** [comment]
Rebased without further changes to fix conflict in util/libcrypto.num.

**Comment by mspncp** [comment]
> That's why backporting this change to 1.1.1 (as a separate pull request) is my primary target. Since the backport is based on #9347, it will depend on the OMC decision about the constification policy.

Ideally, I would like to merge this pull request together with its 1.1.1 backport. The latter however depends on #9347. Is there any upate on #9347?

**Comment by mspncp** [comment]
There will be no cherry-pick to 1.1.1, since  #9347 has been rejected.

**Comment by mspncp** [comment]
Ups, did not mean to close the entire pull request.

**Comment by mspncp** [comment]
Rebased and autosquashed without further changes to resolve conflict in libcrypto.num.

**Comment by mspncp** [comment]
Merged to master, thank you!

a8f1aabd4b x509: publish X509_PUBKEY_dup
9b977675ad x509: add missing X509 dup functions
227d426554 x509: sort X509 dup functions alphabetically

**Pull request closed without merging**